File: InterviewPrep/medium/33_searchInSortedArray.go
/**
 *      file    :  33_searchInSortedArray.go
 *      author  :  techiedheeraj
 *      created :  2020 Nov 21 02:37:19
 *      lastMod :  Sat Nov 21 02:37:19 2020
**/

package main

import (
  "fmt"
)

func search(nums []int, target int) int {

  var l int = 0
  var r int = len(nums)-1

  for l < r {
    var mid int = (l+r)/2

    if nums[mid] > nums[r] {
      l = mid+1
    } else {
      r = mid
    }
  }

  var rot int = l
  l = 0
  r = len(nums)-1

  for l <= r  {
    var mid int = (l+r)/2
    var realmid int = (mid+rot) % len(nums)

    if nums[realmid] == target {
      return realmid
    }
    if nums[realmid] < target {
      l = mid+1
    } else {
      r = mid-1
    }
  }

  return -1

}


/*func search(nums []int, target int) int {

  var hm = make(map[int]int)

  for i, j := range nums {
    hm[j] = i
  }

  k, ok := hm[target]
  if ok == true {
    return k
  }

  return -1
}
*/

File: InterviewPrep/medium/33_searchInSortedArray.cpp
/**
 *      file    :  33_searchInSortedArray.cpp
 *      author  :  techiedheeraj
 *      created :  2020 Nov 21 02:37:49
 *      lastMod :  Sat Nov 21 02:37:49 2020
**/

class Solution {
public:
    int search(vector<int>& nums, int target) {

      int l = 0, r=nums.size()-1;

      while (l < r) {
        int mid = (l+r)/2;

        if (nums[mid] > nums[r]) l = mid+1;
        else r = mid;
      }

      int rot = l;
      l = 0, r = nums.size()-1;

      while (l <= r) {
        int mid = (l+r)/2;
        int realmid = (mid+rot) % nums.size();

        if (nums[realmid] == target)
          return realmid;
        if (nums[realmid] < target) l = mid+1;
        else r = mid-1;
      }
      return -1;

      /*
      unordered_set<int,int>hm;
      for(auto &i: nums)
        hm[nums[i]] = ;

      return hashSet.find(target) != hashSet.end();
      */
    }
};
